recipes-bsp/u-boot/files/0028-iot2050-Add-support-for-configuring-M.2-connector.patch

```diff
@@ -0,0 +1,534 @@
++				    m2_config_table[n].config_pins.config,
++				    sizeof(config_pins.config))) {
++				connector_mode = m2_config_table[n].mode;
++				break;
++			}
++		}
++		if (connector_mode == CONNECTOR_MODE_INVALID) {
++			mode_info = " [fallback, card unknown/unsupported]";
++			connector_mode = BKEY_USB30_EKEY_PCIE;
++		}
++	}
++
++	printf("M.2:   %s%s\n", m2_connector_mode_name[connector_mode],
++	       mode_info);
++
++	/* configure serdes mux */
++	set_pinvalue(serdes_mux_ctl_pin_info[0].gpio_name,
++		     serdes_mux_ctl_pin_info[0].label,
++		     serdes_mux_ctrl[connector_mode].ctrl_usb30_pcie0_lane0);
++	set_pinvalue(serdes_mux_ctl_pin_info[1].gpio_name,
++		     serdes_mux_ctl_pin_info[1].label,
++		     serdes_mux_ctrl[connector_mode].ctrl_pcie1_pcie0);
++	set_pinvalue(serdes_mux_ctl_pin_info[2].gpio_name,
++		     serdes_mux_ctl_pin_info[2].label,
++		     serdes_mux_ctrl[connector_mode].ctrl_usb30_pcie0_lane1);
++
++	m2_overlay_prepare();
++}
++
+ int board_init(void)
+ {
+ 	return 0;
+@@ -215,6 +424,9 @@ int board_late_init(void)
+ 	/* change CTRL_MMR register to let serdes0 not output USB3.0 signals. */
+ 	writel(0x3, SERDES0_LANE_SELECT);
+ 
++	if (board_is_m2())
++		m2_connector_setup();
++
+ 	set_board_info_env();
+ 
+ 	/* remove the eMMC if requested via button */
+@@ -226,6 +438,29 @@ int board_late_init(void)
+ }
+ 
+ #if defined(CONFIG_OF_LIBFDT) && defined(CONFIG_OF_BOARD_SETUP)
++static void m2_fdt_fixup(void *blob)
++{
++	void *overlay;
++	int err;
++
++	if (!connector_overlay)
++		return;
++
++	overlay = malloc(connector_overlay_size);
++	if (!overlay)
++		goto fixup_error;
++
++	memcpy(overlay, connector_overlay, connector_overlay_size);
++	err = fdt_overlay_apply_verbose(blob, overlay);
++	free(overlay);
++
++	if (!err)
++		return;
++
++fixup_error:
++	pr_err("Could not apply M.2 device tree overlay\n");
++}
++
+ int ft_board_setup(void *blob, struct bd_info *bd)
+ {
+ 	int ret;
+@@ -237,6 +472,9 @@ int ft_board_setup(void *blob, struct bd_info *bd)
+ 	if (ret)
+ 		pr_err("%s: fixing up msmc ram failed %d\n", __func__, ret);
+ 
++	if (board_is_m2())
++		m2_fdt_fixup(blob);
++
+ 	return ret;
+ }
+ #endif
+diff --git a/doc/board/siemens/iot2050.rst b/doc/board/siemens/iot2050.rst
+index 0ec14ba7b95..34975541060 100644
+--- a/doc/board/siemens/iot2050.rst
++++ b/doc/board/siemens/iot2050.rst
+@@ -141,3 +141,18 @@ Flash signed flash.bin
+ 
+ The signing has happen in-place in flash.bin, thus the flashing procedure
+ described above.
++
++M.2 slot configuration
++----------------------
++
++The M.2 variant of the IOT2050 comes with one B-key and one E-key slot. These
++are configured by U-Boot depending on the detected usage (auto configuration).
++The device tree loaded later on for the OS will then be fixed up according to
++this configuration.
++
++If this auto configuration does not work reliably, the U-Boot environment
++variable "m2_manual_config" can be set to select the mode manually:
++
```

Comment:
better to add: manual config always prior to auto-detect

Reply:
See above.